EIPS/eip-3675.md

```diff
@@ -0,0 +1,247 @@
+* Add verification of the fields noted in [block structure](#block-structure) as their specified constant value, including the enforcement of empty ommers list.
+* Add verification of block validity with respect to the PoS consensus rules, i.e. assert that there **MUST** be a corresponding event for the block to consider it valid. Such events are limited to the following:
+    * A `POS_CONSENSUS_VALIDATED` event considering the block or any of its descendants as valid.
+    * A `POS_CHAINHEAD_SET` event, including the `POS_BLOCK_FINALIZED` form of this event, designating the block or any of its descendants as the head of the canonical chain.
+
+*Note*: If one of the new rules fails then the block **MUST** be invalidated.
+
+*Note*: Validity rules that are not specified in the list above **MUST** remain unchanged.
+
+
+### Block and ommer rewards
+
+Beginning with `TRANSITION_BLOCK`, block and ommer rewards are deprecated. Specifically, the following actions **MUST** be taken:
+* Remove increasing the balance of the block's **`beneficiary`** account by the basic block reward.
+* Remove increasing the balance of the block's **`beneficiary`** account by the ommer inclusion reward per each ommer.
+* Remove increasing the balance of the ommer's **`beneficiary`** account by the ommer block reward per each ommer.
+
+*Note*: Transaction fee mechanics affecting the block's `beneficiary` account **MUST** remain unchanged.
+
+
+### Fork choice rule
+
+As of the first `POS_CHAINHEAD_SET` event, the fork choice rule **MUST** be altered in the following way: 
+* Remove the existing PoW heaviest chain rule.
+* Adhere to the new PoS LMD-GHOST rule. 
+
+The new PoS LMD-GHOST fork choice rule is specified as follows. On each occurrence of a `POS_CHAINHEAD_SET` event including the first one, the following actions **MUST** be taken:
+* Consider the chain starting at genesis and ending with the block nominated by the event as the canonical blockchain.
+* Set the head of the canonical blockchain to the block nominated by the event.
+
+
+### Network
+
+The networking stack **SHOULD NOT** send the following messages if they advertise the descendant of any terminal PoW block:
+* [`NewBlockHashes (0x01)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblockhashes-0x01)
+* [`NewBlock (0x07)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblock-0x07)
+
+Beginning with the first `POS_BLOCK_FINALIZED` event, the networking stack **MUST** remove the handlers corresponding to the following ETH protocol messages:
+* [`NewBlockHashes (0x01)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblockhashes-0x01)
+* [`NewBlock (0x07)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblock-0x07)
+
+Beginning with the first `POS_BLOCK_FINALIZED` event, peers that keep sending these messages **SHOULD** be disconnected.
```

Comment:
I think this will introduce a race condition.  Not all peers will receive `POS_BLOCK_FINALIZED` at the same time, so some may still be broadcasting `NewBlockHashes` and `NewBlock` for siblings of the terminal PoW block after some of their peers have received the `POS_BLOCK_FINALIZED` event.

Consider changing this rule to be that after the *second* `POS_BLOCK_FINALIZED` event peers are disconnected, as by that time all peers should have received the first `POS_BLOCK_FINALIZED` event and should no longer be gossping sibling terminal PoW blocks.

Reply:
Fair, but there is also the following statement above:
- The networking stack **SHOULD NOT** send the following messages if they advertise the descendant of any terminal PoW block

Which means that effectively block gossip **SHOULD** be stopped before `POS_BLOCK_FINALIZED`.

Reply:
Why would block gossip stop before `POS_BLOCK_FINALIZED`?  I think we should assume that there will be a steady stream of new siblings until `POS_BLOCK_FINALIZED`, since the protocol rules as I understand them suggest you continue to gossip those.

Reply:
Because there is enough time between the terminal PoW block event and the first `POS_BLOCK_FINALIZED` event that is issued when the first PoS block gets finalized (it is not issued when the beacon block with empty execution payload is finalized). The gap between these two events is at least 2 epochs and unless there is a party that is constantly mining terminal PoW blocks there should be a plenty of time to disseminate all terminal PoW blocks existing in the network.

Reply:
Ah, I feel like we should *assume* that one or more miners mine terminal PoW siblings.  They have a huge pile of hashing power and nothing to do with it.  Maybe we'll get lucky and they don't, but I feel like we should prepare for them doing this.

I wonder if a miner could do something interesting where once the first terminal PoW block is produced (the most likely one to be finalized) they switch to mining a forked chain some number of blocks back that lowers the difficulty with each block.  The goal here would be to drive the difficulty down as far as possible in this fork given their hashpower while still reaching the target total difficulty.  Once you get to one block shy of target total difficulty, you then produce terminal PoW blocks over and over as fast as you can in order to spam the network?

I'm not sure how much you could manipulate given 2 epochs.

Reply:
> I'm not sure how much you could manipulate given 2 epochs.

Good question. There are possibilities depending on whether whole hashing power participates in anything like that, whether this attack started in advance by switching a portion of hashing power from mainnet to start producing such a chain before the transition. If all hashing power participates it may produce roughly `60` blocks during 2 epochs period but with lowering the difficulty this number increases; like 20 blocks with 1000 seconds gap between their timestamps is required to reduce the difficulty ~8 times and produce roughly `40*8=320` blocks during the rest of the period, it also doesn't count how many blocks needed to reach the terminal total difficulty threshold. If we assume that somehow miners become able to produce a block in a second then during two epochs we would got `768` terminal PoW blocks, I don't know how bad is it for the network if we assume that these blocks are not executed, only propagated with PoW verification.

As for the network stack, if handlers are removed won't peers sending unknown messages be disconnected by clients?

Reply:
For the health of the network under this kind of attack it makes sense for clients to stop propagating *any* blocks upon receiving the first `POS_BLOCK_FINALIZED` and remove handlers with the corresponding penalty upon receiving the second `POS_BLOCK_FINALIZED`. So, the updated spec should looks like:
> The networking stack **SHOULD NOT** send the following messages if they advertise the descendant of any terminal PoW block:
> * [`NewBlockHashes (0x01)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblockhashes-0x01)
> * [`NewBlock (0x07)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblock-0x07)
>
> Beginning with the *first* `POS_BLOCK_FINALIZED` event, the networking stack **MUST** discard the following ingress messages:
> * [`NewBlockHashes (0x01)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblockhashes-0x01)
> * [`NewBlock (0x07)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblock-0x07)
> 
> Beginning with the *second* `POS_BLOCK_FINALIZED` event, the networking stack **MUST** remove the handlers corresponding to the following messages:
> * [`NewBlockHashes (0x01)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblockhashes-0x01)
> * [`NewBlock (0x07)`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#newblock-0x07)
>
> Peers that keep sending these messages after the handlers have been removed **SHOULD** be disconnected.

Reply:
We should probably move this conversation to the discussions-to link.  PR comments should be assumed to be transient, especially on merged PRs.

Reply:
done, https://ethereum-magicians.org/t/eip-3675-upgrade-consensus-to-proof-of-stake/6706/2